mlflow/llama_index/__init__.py

```diff
@@ -348,32 +354,158 @@ def _load_pyfunc(path, model_config: Optional[Dict[str, Any]] = None):
+    def _build_engine_method(self) -> Callable:
+        if self.engine_type == _QUERY_ENGINE_NAME:
+            return self.index.as_query_engine(**self.model_config).query
+        elif self.engine_type == _CHAT_ENGINE_NAME:
+            return self.index.as_chat_engine(**self.model_config).chat
+        elif self.engine_type == _RETRIEVER_ENGINE_NAME:
+            return self.index.as_retriever(**self.model_config).retrieve
+        else:
+            raise ValueError(
+                f"Unsupported engine type: {self.engine_type}. It must be one of "
+                "{_SUPPORTED_ENGINES}"
+            )
+
+    def __init__(self, index, engine_type: str, model_config: Optional[Dict[str, Any]] = None):
+        self.index = index
+        self.model_config = model_config or {}
+        self.engine_type = engine_type
+        self.predict_callable = self._build_engine_method()
+
+    def _format_predict_input_query_and_retriever(self, data):
+        """Convert pyfunc input to a QueryBundle."""
+        data = _convert_llm_input_data(data)
+
+        if isinstance(data, str):
+            return QueryBundle.from_dict({"query_str": data})
+        elif isinstance(data, dict):
+            try:
+                return QueryBundle.from_dict(data)
+            except KeyError:
+                raise ValueError(
+                    "The input dictionary did not have the correct schema. It must support the "
+                    "supported in a llama_index QueryBundle: "
+                    "['query_str', 'image_path', 'custom_embedding_strs', 'embedding']"
+                )
+        elif isinstance(data, (list, np.ndarray)):
+            prediction_list = [self._format_predict_input_query_and_retriever(d) for d in data]
+        elif isinstance(data, pd.DataFrame):
+            prediction_list = [
+                self._format_predict_input_query_and_retriever(d)
+                for d in data.to_dict(orient="records")
+            ]
+        else:
+            raise ValueError(
+                f"Unsupported input type: {type(data)}. It must be one of "
+                "[str, dict, list, numpy.ndarray, pandas.DataFrame]"
+            )
+
+        if len(prediction_list) == 1:
+            return prediction_list[0]
+
+        return prediction_list
+
+    def _safe_convert_to_chat_messages(self, data) -> Union[List[ChatMessage], str]:
+        # NB: Chat messages that have been converted to a string via str() are formatted as
+        # "`role`: `content`", for example "system: You are a helpful bot".
+        if isinstance(data, (list, tuple)) and all(":" in message for message in data):
+            return [ChatMessage.from_str(*(message.split(": ", 1)[::-1])) for message in data]
+
+        return data
+
+    def _format_predict_input_chat(self, data):
+        """
+        Chat engines can have a variety of signatures. The two primary standards are:
+        1. chat(str, Sequence[ChatMessage])
```

Comment:
Q: in this case, the former is the question and the latter is the chat history?

Reply:
Correct